lib/fuser_mqubit.h

```diff
@@ -0,0 +1,1087 @@
+      unsigned max_fused_size, Scratch& scratch, GateF& fgate) {
+    unsigned level = 2 + scratch.count;
+
+    FindLongestGateSequence(max_fused_size, level, scratch, fgate);
+
+    auto longest_seq = scratch.longest_seq;
+
+    if (longest_seq.size() == 1 && scratch.count == 0) {
+      fgate.visited = 1;
+      return;
+    }
+
+    ++scratch.count;
+
+    for (auto p : longest_seq) {
+      p->gate->visited = 900;
+
+      for (auto q : p->qubits) {
+        fgate.qubits.push_back(q);
+        fgate.mask |= uint64_t{1} << q;
+      }
+
+      for (auto l : p->links) {
+        fgate.links.push_back(l);
+      }
+    }
+
+    // Compress links.
+    for (auto& link : fgate.links) {
+      while (link->prev != nullptr && link->prev->val->visited == 900) {
+        link = link->prev;
+      }
+
+      while (link->next != nullptr && link->next->val->visited == 900) {
+        LinkManager::Delete(link->next);
+      }
+    }
+
+    for (auto p : longest_seq) {
+      p->gate->visited = level;
+    }
+
+    if (longest_seq.size() >= 3) {
+      AddGatesFromNext(longest_seq[2]->gate->gates, fgate);
+    }
+
+    if (longest_seq.size() >= 5) {
+      AddGatesFromNext(longest_seq[4]->gate->gates, fgate);
+    }
+
+    if (longest_seq.size() >= 2) {
+      // May call MakeGateSequence recursively.
+      AddGatesFromPrev(max_fused_size, *longest_seq[1]->gate, scratch, fgate);
+    }
+
+    if (longest_seq.size() >= 4) {
+      // May call MakeGateSequence recursively.
+      AddGatesFromPrev(max_fused_size, *longest_seq[3]->gate, scratch, fgate);
+    }
+
+    for (auto p : longest_seq) {
+      p->gate->visited = 999;
+    }
+
+    FuseNext(1, fgate);
+
+    scratch.gates.push_back(&fgate);
+  }
+
+  static void AddGatesFromNext(std::vector<const Gate*>& gates, GateF& fgate) {
+    for (auto gate : gates) {
+      fgate.gates.push_back(gate);
+    }
+  }
+
+  static void AddGatesFromPrev(unsigned max_fused_size, const GateF& pfgate,
+                               Scratch& scratch, GateF& fgate) {
+    for (auto gate : pfgate.gates) {
+        fgate.gates.push_back(gate);
+    }
+
+    for (auto link : pfgate.links) {
+      if (link->prev == nullptr) continue;
+
+      auto pgate = link->prev->val;
+
+      if (pgate->visited == 1) {
+        MakeGateSequence(max_fused_size, scratch, *pgate);
+      }
+    }
+  }
+
+  static void FindLongestGateSequence(
+      unsigned max_fused_size, unsigned level, Scratch& scratch, GateF& fgate) {
+    scratch.data.push_back({&fgate, {}, {}});
+
+    scratch.longest_seq.resize(0);
+    scratch.longest_seq.push_back(&scratch.data.back());
+
+    scratch.stack.resize(0);
+    scratch.stack.push_back(&scratch.data.back());
+
+    unsigned cur_size = fgate.qubits.size();
+    fgate.visited = level;
+
+    unsigned max_size = cur_size;
+
+    GetNextAvailableGates(max_fused_size, cur_size, fgate, nullptr,
+                          scratch.data, scratch.next1);
+
+    for (auto n1 : scratch.next1) {
```

Comment:
These for-loops reuse a lot of the same code. Could this be restructured as a recursive function call without negatively affecting performance?

Reply:
Removed some code duplication.